mitmproxy/utils/signals.py

```diff
@@ -1,54 +1,134 @@
-    def connect(self, receiver: T, sender: Any = blinker.ANY, weak: bool = True) -> T:
-        # allow better typing than blinker
-        return super().connect(receiver, sender, weak)
-
-    async def send(self, *sender, **kwargs) -> list[tuple[Any, Any]]:
-        return [
-            (
-                receiver,
-                await ret if ret is not None and inspect.isawaitable(ret) else ret,
-            )
-            for receiver, ret in super().send(*sender, **kwargs)
-        ]
+import weakref
+from collections.abc import Callable, Awaitable
+from typing import Any, Generic, TypeVar, cast
+
+try:
+    from typing import ParamSpec
+except ImportError:  # pragma: no cover
+    # Python 3.9
+    from typing_extensions import ParamSpec
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def make_weak_ref(obj: Any) -> weakref.ReferenceType:
+    """
+    Like weakref.ref(), but using weakref.WeakMethod for bound methods.
+    """
+    if hasattr(obj, "__self__"):
+        return cast(weakref.ref, weakref.WeakMethod(obj))
+    else:
+        return weakref.ref(obj)
+
+
+# We're running into https://github.com/python/mypy/issues/6073 here,
+# which is why the base class is a mixin and not a generic superclass.
+class _SignalMixin:
+    def __init__(self):
+        self.receivers: list[weakref.ref[Callable]] = []
+
+    def connect(self, receiver: Callable) -> None:
+        """
+        Register a signal receiver.
+
+        The signal will only hold a weak reference to the receiver function.
+        """
+        receiver = make_weak_ref(receiver)
+        self.receivers.append(receiver)
+
+    def disconnect(self, receiver: Callable) -> None:
+        self.receivers = [r for r in self.receivers if r() != receiver]
+
+    def notify(self, *args, **kwargs):
+        cleanup = False
+        for ref in self.receivers:
+            r = ref()
+            if r is not None:
+                yield r(*args, **kwargs)
+            else:
+                cleanup = True
+        if cleanup:
+            self.receivers = [r for r in self.receivers if r() is not None]
+
+
+class _SyncSignal(Generic[P], _SignalMixin):
+    def connect(self, receiver: Callable[P, None]) -> None:
+        assert not asyncio.iscoroutinefunction(receiver)
+        super().connect(receiver)
+
+    def disconnect(self, receiver: Callable[P, None]) -> None:
+        super().disconnect(receiver)
+
+    def send(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        for ret in super().notify(*args, **kwargs):
+            assert ret is None or not inspect.isawaitable(ret)
+
+
+class _AsyncSignal(Generic[P], _SignalMixin):
+    def connect(self, receiver: Callable[P, Awaitable[None] | None]) -> None:
+        super().connect(receiver)
+
+    def disconnect(self, receiver: Callable[P, Awaitable[None] | None]) -> None:
+        super().disconnect(receiver)
+
+    async def send(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        await asyncio.gather(*[
```

Comment:
Depending on how strict we check `SyncSignal`, maybe also add some `assert`s here.